### PR TITLE
Use aggregation pipeline to speed up neighbor queries

### PR DESCRIPTION
**Pull request by tomchop**

### CODE REVIEW

File: core/database.py [hunk 1]
```diff
@@ -342,18 +345,56 @@ def neighbors(self, neighbor_type=""):
...
+            if key in SEARCH_REPLACE:
+                key = SEARCH_REPLACE[key]
+
+            if regex and isinstance(value, basestring):
+                value = {"$regex": value}
+                if ignorecase:
+                    value["$options"] = "i"
+            if isinstance(value, list) and not key.endswith("__in"):
+                key += "__all"
+            result_filters["related."+key] = value
```
Comment (gaelmuller) [inline core/database.py]: 🚎  When you copy/paste code from somewhere else, this ends up being duplicated code 🚎

File: core/database.py [hunk 2]
```diff
@@ -342,18 +345,56 @@ def neighbors(self, neighbor_type=""):
...
+
+        outnodes = self._neighbors_aggregation("out", klass, result_filters, regex, page, rng)
+        innodes = self._neighbors_aggregation("in", klass, result_filters, regex, page, rng)
```
Comment (gaelmuller) [inline core/database.py]: Did not test it but wouldn't this code result in `rng * 2` results instead of `rng` ?

File: core/database.py (hunk 1 above)
Comment (tomchop) [reply core/database.py]: More importantly, the `__` prefix is invalid when throwing things into the aggregation pipeline. Will fix.

File: core/database.py (hunk 2 above)
Comment (tomchop) [reply core/database.py]: Right, this will result in maximum results of `rng*2` lenght.

**Comment by tomchop** [comment]
The code changes significantly when hitting the mongodb aggregation pipeline directly vs. crafting a Mongoengine queryset

File: core/database.py (hunk 1 above)
Comment (tomchop) [reply core/database.py]: The code changes significantly when hitting the mongodb aggregation pipeline directly vs. crafting a Mongoengine queryset

**Comment by tomchop** [comment]
Merging after getting @gaelmuller's offline approval.

**Pull request merged**